hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/EndOfStakingPeriodUpdater.java

```diff
@@ -0,0 +1,412 @@
+        this.accountNumbers = accountNumbers;
+        this.stakeInfoHelper = stakeInfoHelper;
+    }
+
+    /**
+     * Updates all (relevant) staking-related values for all nodes, as well as any network reward information,
+     * at the end of a staking period. This method must be invoked during handling of a transaction
+     *
+     * TODO: this method is still not called by the handle workflow (will be done in issue #7721)
+     *
+     * @param consensusTime the consensus time of the transaction used to end the staking period
+     * @param context the context of the transaction used to end the staking period
+     */
+    public void updateNodes(@NonNull final Instant consensusTime, @NonNull final HandleContext context) {
+        log.info("Updating node stakes for a just-finished period @ {}", consensusTime);
+
+        // First, determine if staking is enabled. If not, there is nothing to do
+        final var stakingConfig = context.configuration().getConfigData(StakingConfig.class);
+        if (!stakingConfig.isEnabled()) {
+            log.info("Staking not enabled, nothing to do");
+            return;
+        }
+
+        final ReadableAccountStore accountStore = context.readableStore(ReadableAccountStore.class);
+        final WritableStakingInfoStore stakingInfoStore = context.writableStore(WritableStakingInfoStore.class);
+        final WritableNetworkStakingRewardsStore stakingRewardsStore =
+                context.writableStore(WritableNetworkStakingRewardsStore.class);
+
+        final var nodeIds = stakingInfoStore.getAll();
+        final var totalStakedRewardStart = stakingRewardsStore.totalStakeRewardStart();
+        final var rewardRate =
+                rewardRateForEndingPeriod(totalStakedRewardStart, accountStore, stakingRewardsStore, stakingConfig);
+        // The tinybars earned per hbar for stakers who were staked to a node whose total stakedRewardStart for the
+        // ending period was in the range [minStake, maxStake]
+        final var perHbarRate = totalStakedRewardStart < HBARS_TO_TINYBARS
+                ? 0
+                : rewardRate / (totalStakedRewardStart / HBARS_TO_TINYBARS);
+        log.info(
+                "The reward rate for the period was {} tb ({} tb/hbar for nodes with in-range stake, given {} total stake reward start)",
+                rewardRate,
+                perHbarRate,
+                totalStakedRewardStart);
+
+        // Calculate the updated stake and reward sum history for each node
+        long newTotalStakedStart = 0L;
+        long newTotalStakedRewardStart = 0L;
+        long maxStakeOfAllNodes = 0L;
+        final Map<Long, StakingNodeInfo> updatedNodeInfos = new HashMap<>();
+        for (final var nodeNum : nodeIds.stream().sorted().toList()) {
+            var currStakingInfo = stakingInfoStore.getForModify(nodeNum);
+
+            // The return value is the reward rate (tinybars-per-hbar-staked-to-reward) that will be paid to all
```

Comment:
Here return value is not just reward rate right ?

Reply:
You're right, fixed this comment